source/slang/slang-ir-lower-generics.cpp

```diff
@@ -0,0 +1,159 @@
+// slang-ir-lower-generics.cpp
+#include "slang-ir-lower-generics.h"
+
+#include "slang-ir.h"
+#include "slang-ir-clone.h"
+#include "slang-ir-insts.h"
+
+namespace Slang
+{
+    struct GenericsLoweringContext;
+
+    struct GenericsLoweringContext
+    {
+        // For convenience, we will keep a pointer to the module
+        // we are specializing.
+        IRModule* module;
+
+        Dictionary<IRInst*, IRInst*> loweredGenericFunctions;
+
+        SharedIRBuilder sharedBuilderStorage;
+
+        // We will use a single work list of instructions that need
+        // to be considered for specialization or simplification,
+        // whether generic, existential, etc.
+        //
+        List<IRInst*> workList;
+        HashSet<IRInst*> workListSet;
+
+        void addToWorkList(
+            IRInst* inst)
+        {
+            // We will ignore any code that is nested under a generic,
+            // because it doesn't make sense to perform specialization
+            // on such code.
+            //
+            for (auto ii = inst->getParent(); ii; ii = ii->getParent())
+            {
+                if (as<IRGeneric>(ii))
+                    return;
+            }
+
+            if (workListSet.Contains(inst))
+                return;
+
+            workList.add(inst);
+            workListSet.Add(inst);
+        }
+
+        IRInst* lowerGenericFunction(IRInst* genericValue)
+        {
+            IRInst* result = nullptr;
+            if (loweredGenericFunctions.TryGetValue(genericValue, result))
+                return result;
+            auto genericParent = as<IRGeneric>(genericValue);
+            SLANG_ASSERT(genericParent);
+            auto func = as<IRFunc>(findGenericReturnVal(genericParent));
+            SLANG_ASSERT(func);
+            if (!func->isDefinition())
+            {
+                loweredGenericFunctions[genericValue] = genericValue;
+                return genericValue;
+            }
```

Comment:
It seems like we would eventually want to support the case where the generic function is an imported declaration rather than a definition, so that we can lower its signature as expected and then rewrite call sites to invoke it as expected.

I guess that won't work in the short term where we have a *lot* of built-in functions that we really need to invoke as generics with implicit full specialization using the templates in the "prelude" file. Trying to invoke those things as dynamic-dispatch generics today would require massive changes to the prelude file and completely tank the performance of our CPU code.